### PR TITLE
Feature/pallet encrypted transactions

### PR DESCRIPTION
**Pull request by iStrike7**
Added pallet for encrypted transactions.

No api breaking changes.

### CODE REVIEW

File: pallets/encrypted-transactions/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+        if let Some(previous_session_index) = session_index.checked_sub(1u8.into()) {
+            TxnRecord::<T>::remove_prefix(T::Index::from(previous_session_index));
+            ExecutedTxnRecord::<T>::remove_prefix(T::Index::from(previous_session_index));
+        }
+    }
+
+    fn on_disabled(_i: usize) {
+        // ignore
+    }
+}
```
Comment (mateuszaaa) [inline pallets/encrypted-transactions/src/lib.rs:274]: some unit testing would be nice- please create proper task for it

**Review by mateuszaaa — CHANGES_REQUESTED** [review]

File: pallets/encrypted-transactions/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+serde = { version = "1.0.101", optional = true }
+pallet-session = { version = "2.0.0", default-features = false, git = "https://github.com/mangata-finance/substrate", branch = "mangata-dev" }
+sp-io = { version = "2.0.0", default-features = false, git = "https://github.com/mangata-finance/substrate", branch = "mangata-dev" }
+sp-runtime = { version = "2.0.0", default-features = false, git = "https://github.com/mangata-finance/substrate", branch = "mangata-dev" }
+frame-support = { version = "2.0.0", default-features = false, git = "https://github.com/mangata-finance/substrate", branch = "mangata-dev" }
+frame-system = { version = "2.0.0", default-features = false, git = "https://github.com/mangata-finance/substrate", branch = "mangata-dev" }
+mangata-primitives = { default-features = false, version = '0.1.0' , path = '../../primitives/mangata'}
+frame-benchmarking = { version = "2.0.0", default-features = false, git = "https://github.com/mangata-finance/substrate", branch = "mangata-dev" , optional = true }
+orml-tokens = {  default-features = false, version = '0.3.1' , git = "https://github.com/mangata-finance/open-runtime-module-library", branch = "mangata-dev" }
```
[the remark this answers is not in the document]
Comment (iStrike7) [reply pallets/encrypted-transactions/Cargo.toml:23]: We do need all of these, they are all imported.

File: pallets/encrypted-transactions/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+sp-std = { version = "2.0.0", default-features = false, git = "https://github.com/mangata-finance/substrate", branch = "mangata-dev" }
+serde = { version = "1.0.101", optional = true }
+pallet-session = { version = "2.0.0", default-features = false, git = "https://github.com/mangata-finance/substrate", branch = "mangata-dev" }
+sp-io = { version = "2.0.0", default-features = false, git = "https://github.com/mangata-finance/substrate", branch = "mangata-dev" }
+sp-runtime = { version = "2.0.0", default-features = false, git = "https://github.com/mangata-finance/substrate", branch = "mangata-dev" }
+frame-support = { version = "2.0.0", default-features = false, git = "https://github.com/mangata-finance/substrate", branch = "mangata-dev" }
+frame-system = { version = "2.0.0", default-features = false, git = "https://github.com/mangata-finance/substrate", branch = "mangata-dev" }
+mangata-primitives = { default-features = false, version = '0.1.0' , path = '../../primitives/mangata'}
+frame-benchmarking = { version = "2.0.0", default-features = false, git = "https://github.com/mangata-finance/substrate", branch = "mangata-dev" , optional = true }
```
[the remark this answers is not in the document]
Comment (iStrike7) [reply pallets/encrypted-transactions/Cargo.toml:22]: Are you talking about frame-benchmarking? Yeah, we can remove that for now.

File: node/Cargo.toml [hunk 4]
```diff
@@ -64,6 +64,7 @@ pallet-authority-discovery = { version = '2.0.0', git = "https://github.com/mang
 pallet-grandpa = { version = '2.0.0', git = "https://github.com/mangata-finance/substrate", branch = "mangata-dev" }
 
 pallet-staking = {version='2.0.0', path="../pallets/staking"}
+pallet-encrypted-transactions = {version='0.1.0', path="../pallets/encrypted-transactions"}
```
[the remark this answers is not in the document]
Comment (iStrike7) [reply node/Cargo.toml:67]: Yeah, pallet-encrypted-tx should be removed. Perhaps separate PR?

File: pallets/encrypted-transactions/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+
+decl_storage! {
+    trait Store for Module<T: Trait> as EncryptedTransactions {
+
+        /// The current set of keys that may issue a heartbeat.
+        KeyMap get(fn keys): BTreeMap<T::AccountId, T::AuthorityId>;
+
+        /// The registry for all the user submitted transactions
+        TxnRegistry get(fn txn_registry): map hasher(blake2_128_concat) T::Hash => (Vec<u8>, T::AccountId, T::Index, Weight, T::AccountId, T::AccountId, Option<Vec<u8>>);
```
[the remark this answers is not in the document]
Comment (iStrike7) [reply pallets/encrypted-transactions/src/lib.rs:68]: Do you mean a use a tuple struct? Would that make much difference?

File: pallets/encrypted-transactions/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+    }
+);
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+
+        fn deposit_event() = default;
+
+        #[weight = 10_000]
+        fn submit_doubly_encrypted_transaction(origin, doubly_encrypted_call: Vec<u8>, nonce: T::Index, weight: Weight, builder: T::AccountId, executor: T::AccountId) -> DispatchResult{
```
[the remark this answers is not in the document]
Comment (iStrike7) [reply pallets/encrypted-transactions/src/lib.rs:129]: Hmm, yes we should do that. But do we understand what that limit should be?

File: pallets/encrypted-transactions/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+            TxnRecord::<T>::mutate(T::Index::from(<pallet_session::Module<T>>::current_index()), &user, |tree_record| tree_record.insert(identifier, (fee_charged, false)));
+            Self::deposit_event(RawEvent::DoublyEncryptedTxnSubmitted(user, nonce));
+            Ok(())
+        }
+
+        #[weight = 10_000]
+        fn submit_singly_encrypted_transaction(origin, identifier: T::Hash, singly_encrypted_call: Vec<u8>) -> DispatchResult{
+            ensure_none(origin)?;
+            let (doubly_encrypted_call, user, nonce, weight, builder, executor, _) = TxnRegistry::<T>::get(identifier);
+            DoublyEncryptedQueue::<T>::mutate(&builder, |vec_hash| {vec_hash.retain(|x| *x!=identifier)});
```
[the remark this answers is not in the document]
Comment (iStrike7) [reply pallets/encrypted-transactions/src/lib.rs:156]: 1. Use the client for that? As we are already doing quite a bit of storage access and computation on the client side, pushing this onto the slower runtime with the same result (rejecting the block), doesn't seem the best approach.
2. I thought about this, and chose to keep it flexible for now as we do not have scheduling and weight considerations sorted out. So maybe for the testnet till we move to Aura we could keep it generic.

File: pallets/encrypted-transactions/src/lib.rs (hunk 1 above)
Comment (iStrike7) [reply pallets/encrypted-transactions/src/lib.rs:274]: https://trello.com/c/Wzj7gb5w/312-unit-testing-for-pallet-encrypted-transactions

File: pallets/encrypted-transactions/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+            SinglyEncryptedQueue::<T>::mutate(&executor, |vec_hash| {vec_hash.push(identifier)});
+            TxnRegistry::<T>::insert(identifier, (doubly_encrypted_call, user, nonce, weight, &builder, &executor, Some(singly_encrypted_call)));
```
[the remark this answers is not in the document]
Comment (iStrike7) [reply pallets/encrypted-transactions/src/lib.rs:158]: I don't think that can be done.

File: pallets/encrypted-transactions/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mateuszaaa) [reply pallets/encrypted-transactions/src/lib.rs:68]: i was rather thinking about regular struct

File: pallets/encrypted-transactions/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mateuszaaa) [reply pallets/encrypted-transactions/src/lib.rs:129]: i think we can calculate that at least for XYK pallet for now ?

[the remark this answers is not in the document]
Comment (mateuszaaa) [reply pallets/encrypted-transactions/src/lib.rs:129]:
```
sizeof(2x TokenID) + sizeof(2xBalance) + sizeof (origin) + sizeof(pallet id) + sizeof(extrinsic id)
```

so something like
```
2*32 + 2*128 + 512(?) + 512(probably its less) + 512(probably its less)
```
and we can double the value just for now

File: pallets/encrypted-transactions/src/lib.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (mateuszaaa) [reply pallets/encrypted-transactions/src/lib.rs:158]: can you elaborate a bit more ?

File: node/Cargo.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (mateuszaaa) [reply node/Cargo.toml:67]: please create task for it

[the remark this answers is not in the document]
Comment (iStrike7) [reply node/Cargo.toml:67]: Added a commit removing it.

File: pallets/encrypted-transactions/Cargo.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (iStrike7) [reply pallets/encrypted-transactions/Cargo.toml]: Removed.

File: pallets/encrypted-transactions/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (iStrike7) [reply pallets/encrypted-transactions/src/lib.rs]: Now using struct TxnRegistryDetails

File: pallets/encrypted-transactions/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (iStrike7) [reply pallets/encrypted-transactions/src/lib.rs:148]: Using DoublyEncryptedCallMaxLength as 4096 bytes.

File: pallets/encrypted-transactions/src/lib.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (iStrike7) [reply pallets/encrypted-transactions/src/lib.rs]: Improved the implementation

[the remark this answers is not in the document]
Comment (iStrike7) [reply pallets/encrypted-transactions/src/lib.rs]: Improved the implementation.

**Review by mateuszaaa — APPROVED** [review]
LGTM

File: pallets/encrypted-transactions/src/lib.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (iStrike7) [reply pallets/encrypted-transactions/src/lib.rs]: > can you elaborate a bit more ?
You can't really just update one field of a stored item. You will still have to read the while thing and then write the whole thing.
The improved implementation however reduces 1 storage read.

**Pull request merged**